### ISSUE TITLE
Consider removing serverCertificateHashes

### ISSUE THREAD

**Issue by annevk** [w3c/webtransport#623]
Without more restrictions this feature could lead to the same kind of problem plaguing Web Push, where messages are not end-to-end encrypted because it's easier to let middleware worry about the complexity of encryption and dealing with the Web Browser Endpoint in general.

I also heard that the main motivation for this feature is local development and testing, which does not meet the high bar we have for web platform features.

As things stand today WebKit does not intend to implement this.

**Comment by vasilvv** [comment]
> Without more restrictions this feature could lead to the same kind of problem plaguing Web Push, where messages are not end-to-end encrypted because it's easier to let middleware worry about the complexity of encryption and dealing with the Web Browser Endpoint in general.

I do not understand this comparison.  What does Web Push or middleware has to do with serverCertificateHashes?

> I also heard that the main motivation for this feature is local development and testing, which does not meet the high bar we have for web platform features.

This is not correct; the motivation for this feature is to provide the ability to connect to endpoints that cannot have a publicly trusted certificate, in order to have feature parity with WebRTC in that regard.  The original reasoning for adding it is documented in #18.  Since that discussion happened, more motivation came from scenarios like WICG/private-network-access#23.

**Comment by annevk** [comment]
> I do not understand this comparison. What does Web Push or middleware has to do with serverCertificateHashes?

The risk that people will just use a shared private certificate across a range of devices just to get this to work without actually caring much about end user security.

And while there is definitely a need for some kind of HTTPS for local networks, we should not solve it on a per-API basis.

**Comment by youennf** [comment]
> in order to have feature parity with WebRTC in that regard

WebRTC is peer-to-peer hence the current solution for WebRTC.
WebTransport being client/server, it seems more consistent to stick with what other client/server networking stacks like HTTP or WebSocket are doing.

**Comment by vasilvv** [comment]
> The risk that people will just use a shared private certificate across a range of devices just to get this to work without actually caring much about end user security.

We have discussed this issue previously in #349; this resulted in an expiry requirement being added, as documented in <https://w3c.github.io/webtransport/#certificate-hashes>.

**Comment by annevk** [comment]
I don't see how that addresses the concern. You can generate fresh certificates with a shared private key quite easily, but it's not actually secure.

**Comment by vasilvv** [comment]
> You can generate fresh certificates with a shared private key quite easily, but it's not actually secure.

I don't believe this is in any way different from regular Web PKI certificates.

**Comment by jan-ivar** [comment]
Meeting:
- Write down all the reasons this is ok
- @jesup to check Mozilla position
- See also https://github.com/mozilla/standards-positions/issues/167#issuecomment-1015542275

**Comment by annevk** [comment]
@vasilvv regular certificates have all kinds of safeguards, including being able to distrust the root if practices like this were to be discovered.

**Comment by vasilvv** [comment]
As far as I am aware, there isn't anything in CA/B BR that requires CAs to check for private key reuse; furthermore, as far as I am aware, reusing private keys is not uncommon in practice.

**Comment by martenrichter** [comment]
Just a few thoughts: I use the hashes for:
a) Development, as webtransport has no HTTP. I was not able to set up a development setup without it.
b) For VMs rented from cloud providers. Without this mechanism, one must at least integrate them into a DNS domain for let's encrypt to work. Or set up some automated CA issuer. All this requires a much more infrastructure. (Currently I have to do both because of the websocket fallback).

Regarding reusing certificates:
c.) Yes, people do stupid things, but it would always be bound to the JS code of a particular website. So, if this is compromised, everything is already lost.
d.) One may decide that the IP  (for the VM use case) or hostname is set on the CN field and is checked (in the moment 0.0.0.0) and forbid wildcards. In this way, mechanisms are in place to generate certificates for each client individually.

Support in Firefox:
e) In light of this discussion, did Mozilla break  serverCertificateHashes intentionally (It requires a root CA in the moment):
https://bugzilla.mozilla.org/show_bug.cgi?id=1934402
f) If no, does it make sense to patch it?
g) Would it make sense to invest time to provide time for a wpt test to prevent such breakage? Or may it happen fast that the feature is removed?

Of course, I would prefer that Webkit support the hashes. But for a web app it may also be an option to recommend other browsers if European DMA results in Gecko and Blink appearing on iOS etc.

**Comment by martenrichter** [comment]
Regarding e-g) I have made it anyway: https://phabricator.services.mozilla.com/D231479 .

**Comment by martinthomson** [comment]
> messages are not end-to-end encrypted 

I want to address this, because it is a misconception about this feature.

Consider the two cases where a site, "https://some.example" say, initiates a WebTransport connection:

1. The site specifies a domain name: "https://example.com".
2. The site specifies a certificate hash (some long hex string).

In neither case does the identity of the WebTransport host come from something that the user specifies.  The root of trust is the site.

In both cases, an attack on the site that results in providing the wrong server identity causes the connection to be vulnerable to interception.

The only difference that I've been able to identify is that the identity of the server in the first case will be logged in a CT log somewhere.

> The risk that people will just use a shared private certificate across a range of devices just to get this to work 

The identified remedy is only capable of making the API a modest amount harder to use.  It doesn't change the security posture meaningfully.  Servers like Caddy or tools like [mkcert](https://github.com/FiloSottile/mkcert) largely eliminate those differences.  Of course, the use of the WebPKI tends to lead to people sharing the same private key across a range of devices as well, so it's a bit meaningless.

This feature exists in support of rapid deployment, which includes development scenarios.  While I'm somewhat ambivalent about its general utility in a world that includes widespread ACME deployment, I don't think that there's a meaningful security argument for its removal.

**Comment by vasilvv** [comment]
> This feature exists in support of rapid deployment, which includes development scenarios. While I'm somewhat ambivalent about its general utility in a world that includes widespread ACME deployment, I don't think that there's a meaningful security argument for its removal.

I feel like local network access use cases are not really addressed by ACME, since the server in question would normally not be accessible by a public CA.  Right now, I'm aware of four options available to web developers that need to do something like that:

1. Get a public CA to issue a certificate to a name that resolves to a private IP, and ship that certificate to the device in question (does not work consistently due to DNS rebinding protection, though I heard there are some ways around that that).
2. Use WebRTC.
3. Use WebTransport with serverCertificateHashes.
4. Give up and use plain HTTP.
5. Wait until the new API that solves this specific problem materializes (note that this does not count as an option).

Out of those four options, only 2 and 4 work consistently in all web browsers.  2 has proven sufficiently unpalatable to web developers that they have been willing to settle for 4; this situation does not appear to be desirable state for the ecosystem.  I believe that 3 could be a sufficient improvement over 2 (in terms of both developer experience, and safeguards against misuse) that it might steer developers away from 4.

> And while there is definitely a need for some kind of HTTPS for local networks, we should not solve it on a per-API basis.

Solving this problem for all forms of HTTP requests in general seems to me like boiling the ocean, mostly due to challenges unique to HTTP (caching, top level navigation, cookies, etc).  Solving this on per-API basis is more like boiling a water in a kettle, in the sense that we do know how to do that, and there are at least two implemented APIs (one of which is shipped in every major browser engine) that already do that.

**Comment by annevk** [comment]
> But for a web app it may also be an option to recommend other browsers if European DMA results in Gecko and Blink appearing on iOS etc.

This does not seem like acceptable discourse.

**Comment by martenrichter** [comment]
> > But for a web app it may also be an option to recommend other browsers if European DMA results in Gecko and Blink appearing on iOS etc.
> 
> This does not seem like acceptable discourse.

Why? It is not insulting, but European legislation about market regulation. It is my reasoning for a small open-source project with no capacity for developing fallbacks for the browsers that do not have newer apis for supporting some critical features (I polyfil for some, but... e.g. webtransport over websocket polyfil, and I have also wrote patches for some browsers).  I recommend using the browser that supports the newer features. Only on ios, there is lack of competition, and I would like to support these devices as well fully without much extra work, the option of a switch of browser engine would solve this problem, of course I would prefer to have features on all three engines.... as I am solution oriented, I accept any solution. And serverCertificateHashes is a feature, that I believe is secure and easy to deploy and this ideal for smaller projects.

**Comment by mozfreddyb** [comment]
@martenrichter Your point is off-topic for this conversation. Let's focus on the technical merits of the proposal and the pros & cons of using WebTransport and its alternatives.

**Comment by jan-ivar** [comment]
Meeting:
- example of user need: https://bugzilla.mozilla.org/show_bug.cgi?id=1806693#c1
- revisit next meeting

**Comment by jan-ivar** [comment]
Meeting:
- What are production use cases outside of development debugging?
-  Talk to device on local network, can't get publicly trusted cert.
- If the device has some sort of other trust arrangement with a server that does have a cert, could it just issue a cert?
- A lot of routers will disallow you to DNS to the server you want (?)
- Martin: not seeing a security issue here
- Number of certificates needed might be high
- We lose some of the transparency PKI would provide
- We think this is not opening a security hole because (1) if it's there it's also in WebRTC, and (2) the abuses discussed in this issue are generally also possible with PKI (how is an ephemeral certificate with a real name different?)
- One difference might be, with a real cert, if we discover people doing iffy things with it, we can revoke it
- For parity with WebRTC, in the connect to local device use case, transparency might not be desired, and be less private
- Pit this against the ability to revoke certs from these people against their privacy?
- Iffy things: ship a device to lots of people, just take a particular key pair, generate a cert, bake it in the device and never update it. Mitigated by existing expiry date
- Is this worse than with PKI wrt people being able to reuse keys?

**Comment by kixelated** [comment]
I currently use `serverCertificateHashes` but only because Chrome doesn't seem to work with [mkcert](https://github.com/FiloSottile/mkcert) (HTTPS works but not WebTransport). I would be okay with removing support once that implementation detail has been fixed. A few thoughts:

1) WebRTC needs certificate hashes due to its peer-to-peer design. Peers cannot use DNS or system signed roots to validate other peers. This is not the case with WebTransport (or HTTPS or WebSockets) given the client-to-server design. Copying HTTPS makes sense.

2) Certs accepted via `serverCertificateHashes` are valid for 2 weeks at most. Any long-running server has to figure out a certificate rotation (and revocation?) scheme via a secure mechanism otherwise you're vulnerable to MitM. Using HTTPS to fetch cert hashes is not an option as that requires cert using a system root, defeating the point of using `serverCertificateHashes` in the first place. 

3) I agree that `serverCertificateHashes` is not a security vulnerability when used correctly. However, the ephemeral nature of the certs encourages insecure and custom methods of procuring the cert hashes. For example, I'm currently using HTTP to insecurely fetch the certificate hash (fortunately only over localhost). But maybe some developer will get clever and procedurally generate private certificates and their corresponding hash...

4) Removing the 2 week expiration would allow baking public keys into binaries, for better or for worse. Revokation would not be supported but the same can be said about a self-signed root CA. It might be useful for local networks or applications with secure upgrade paths (to revoke compromised certs).

5) TLS is definitely annoying for local development. Back when I used mkcert for local development, I would often get bug reports when users tried to run the client and server on different hosts or in docker. It's not obvious that you have to generate a self-signed root CA and certificate on the _client_ and then transfer it to the server. I spent most of my time trying to explain how certificates work, not how WebTransport works.

The current approach feels like a work-around and not a solution. I don't care either way if `serverCertificateHashes` gets removed.

**Comment by marten-seemann** [comment]
libp2p uses `serverCertificateHashes`. At my previous job, I wrote a detailed blog post about this: https://blog.libp2p.io/2022-12-19-libp2p-webtransport/

Removing `serverCertificateHashes` would break this use case, and make WebTransport a lot less useful for the p2p (or specific: the browser to full node) use case.

**Comment by vasilvv** [comment]
> I currently use serverCertificateHashes but only because Chrome didn't seem to work with [mkcert](https://github.com/FiloSottile/mkcert) (HTTPS worked but not WebTransport).

Chrome works with mkcert as long as you have dev mode in `chrome://flags` enabled.  That said, making local development easier was never the goal for `serverCertificateHashes` (the main target was supporting ephemeral VMs and local network connections).

**Comment by martenrichter** [comment]
> * A lot of routers will disallow you to DNS to the server you want (?)

If it is a VM from a cloud provider, you can just use the DNS name assigned from the cloud. At least, this is the way I have it in my current installation. (Open Telekom Cloud). So this is not an issue. The bigger problem is that UDP is blocked in a lot of networks(university) and I have to fall back to websocket, but this is another story.

>  * If the device has some sort of other trust arrangement with a server that does have a cert, could it just issue a cert?
For my audio video routers, they share a secret with a main controlling server, generate a JSON web token, deposit the hash and DNS name at the server, generate the certificate themselves, and also automatically rotate the certificate. (It is all on GitHub.) It is very stable; I have not had to fix anything for over a year now. But the same is true for the Let's Encrypt certificate I need for WebSockets.

>  * One difference might be, with a real cert, if we discover people doing iffy things with it, we can revoke it
But in order to ramp up a WebTransport connection, one has to be in a secure context, so one would revoke this certificate instead, or have a list of revoked hashes?

> * Iffy things: ship a device to lots of people, just take a particular key pair, generate a cert, bake it in the device and never update it. Mitigated by existing expiry date

Only a device might still bake a private key into the device and generate certificates dynamically.

> Chrome works with mkcert as long as you have dev mode in chrome://flags enabled. That said, making local development easier was never the goal for serverCertificateHashes (the main target was supporting ephemeral VMs and local network connections).

Okay, interesting. Actually, when I started developing web transport code, I never got it working with local certificates and Chrome. So servercerticateHashes came to the rescue. (After spending days trying it, where probably a lot of people would use WebSockets and WebRTC out of frustration).
If you start setting up a dev environment, it is a big hurdle with TLS for local development.
Here, it would be good to have something like a how-to, which I may, for example, include in the readme of my node.js package. And maybe others providing similar packages, 


Besides, I think a very important point against the stupid use of serverCertificateHashes is education.
Providing example codes and best practices in articles can increase the security of this feature. There is not much material out there, and I think the first good article (I do not mean the spec; the style is a bit terse) is probably getting copied a lot by humans and AI.
I have at least open source code for certificate generation available for node.js (tests of the node package) and python (pull request against wpt tests infrastructure pending initially on Firefox and now transferred by jgraham to github).

**Comment by kixelated** [comment]
> > I currently use serverCertificateHashes but only because Chrome didn't seem to work with [mkcert](https://github.com/FiloSottile/mkcert) (HTTPS worked but not WebTransport).
> 
> Chrome works with mkcert as long as you have dev mode in `chrome://flags` enabled. That said, making local development easier was never the goal for `serverCertificateHashes` (the main target was supporting ephemeral VMs and local network connections).

Let me get this straight. `chrome://flags` WebTransport Developer Mode reads:
> When enabled, removes the requirement that all certificates used for WebTransport over HTTP/3 are issued by a known certificate root.

Does that:

1. Use system root CAs instead of the Chrome built-in root CAs?
2. or disable TLS certificate validation entirely? (a literal interpretation)
3. or something else?

I think it's (1) based on what you're saying. `mkcert` installs a system root which Chrome happily accepts for HTTPS traffic but not WebTransport without the flag. VPNs often install system roots, so they won't work without Developer Mode either?

But if it's (2) then oh no.

**Comment by wilaw** [comment]
Noting from [minutes ](https://www.w3.org/wiki/WebTransport/Meetings2023#WebTransport_Bi-weekly_Virtual_Meeting_#94_late_-_Jan_28,_2025) of the last call that @ekinnear agree to sync up with internal folk to discuss whether the security risks associated with serverCertificateHashes were no worse than those afforded by existing WebPKI key re(ab)use.

**Comment by jan-ivar** [comment]
Meeting:
- might it be reasonable to limit this to a local network?
- is there any reason this needs to be in version 1 of webtransport?
- still no consensus this is totally safe
- Does anyone have a version of this that still solves our goals?
- unloved because it is not WebPKI
- complicates implementations
- if we acknowledge the use case we need to solve it somehow
- Non-debug use case is peer to peer to end points without webPKI
- Would like to see a clearer explanation of why this is not safe
- @ekinnear to take get another round of feedback, time box: 6 weeks-ish

**Comment by jan-ivar** [comment]
Note we also have https://github.com/w3c/webtransport/issues/59 to solve specifically for serverCertificateHashes. If we''re talking about removing it wholeslate, might another option to leave it off by default (unless requested through CSP?)

**Comment by annevk** [comment]
> if we acknowledge the use case we need to solve it somehow

Why is this true? Surely this is a trade off just like everything else.

**Comment by mozfreddyb** [comment]
A secure page currently can not talk to a local network page even if they intend to co-operate. Mixed content blocking makes it basically impossible for HTTPS to talk to HTTP. Typical domain validation schemes (e.g., ACME) makes it very cumbersome to get an HTTPS certificate for a local network device. A rotating, short-lived certificate hash might be a good trade off for that scenario.

I mean, what else would work?

**Comment by annevk** [comment]
I'm not sure if you're attempting to answer my question or are replying to something else, but I'm specifically asking why, acknowledging the use case means that we have to solve it. It seems perfectly valid to acknowledge the use case and not address it for now as the problem is bigger than just WebTransport, for instance.

**Comment by jan-ivar** [comment]
> > if we acknowledge the use case we need to solve it somehow
> 
> Why is this true? Surely this is a trade off just like everything else.

You are right we can acknowledge the use case without addressing it. My _"Meeting:"_ notes are an attempt to write down what I hear during the meeting, and are not meant to reflect a resolution.

Our more authoritative set of meeting notes are [here](https://www.w3.org/wiki/WebTransport/Meetings2023#WebTransport_Bi-weekly_Virtual_Meeting_#96_late_-_Feb_25,_2025), we also have recordings, but no decision has been made here.

We're looking for a clear explanation of why this is not safe.

**Comment by ekinnear** [comment]
Using the pros and cons list from the slides presented in the latest W3C update to the IETF: https://datatracker.ietf.org/meeting/122/materials/slides-122-webtrans-webtransport-slides-01

### Pros
1. Needed for local network / ephemeral scenarios
See a strong use case for locally hosted devices that cannot easily get publicly valid certificates (e.g. local network cameras, ephemeral cloud VMs). Certificate pinning by hash works around the usual DNS + CA approach.
2. No worse than typical cert usage
 In practice, PKI certificates often reuse private keys, and short-lived certs can mitigate abuse. They argue that “shared private keys” is already possible with publicly trusted certs, so removing hash-based pinning doesn’t eliminate that risk.
3. WebRTC parity
WebRTC (which is peer-to-peer) also uses a certificate-based identity. In principle, it is not so different to let WebTransport connect to a non-publicly trusted endpoint—just a client–server version of that.
4. Developer experience
 People want to avoid forcing developers to jump to WebRTC or unencrypted local HTTP. They say that by removing serverCertificateHashes, you might drive developers to worse security (plain HTTP or complicated setup).

### Cons
1. Security risk / parallels to Web Push
A repeat of the “shared private key” problem, where people distribute the same private key widely (e.g. across many devices) as a quick hack. That undermines real security, because anyone who gets that key can intercept or impersonate the endpoint.
2. Local dev alone does not justify a new API
The original motivation—local testing or development—might not meet a high bar for web platform features.
3. Need a broader, not per‑API, solution
 Since the real problem is local HTTPS or ephemeral cert usage, solving it for just WebTransport may be short-sighted. They suggest the web platform should address local HTTPS in a more unified way.
4. No buy-in from some UA engines
 WebKit says it will not implement it, while Chrome and  Mozilla are in favor of retaining serverCertificateHashes and have implemented support.

**Comment by martinthomson** [comment]
Thanks for doing the homework Eric.

I agree with Cons 2 and 3, but don't agree that either is sufficient cause to block a useful feature.  Con 2 only addresses one use case, not all, with that one case being trivial.  Con 3 is an architectural purity argument being given precedence over other constituencies.

I don't agree that Con 1 is a problem.  Yes, this would depend on having a system for distributing updated hashes, rather than indirecting through a PKI and names.  But assuming that it will lead to increased sharing of secret keys assumes facts not in evidence.  WebRTC uses the same distribution requirement successfully.  While you might not agree that this is less useful when servers cannot run on other browsers, there are plenty of cases in WebRTC where servers use keys that could be distributed using a PKI, but they don't.

This seems to be about not acknowledging the validity of the use case more than anything; I'd like to hear from WebKit about why they don't see that use case as valid.

Finally, it's obvious that our standards are voluntary and Safari can choose not to implement (Con 4).  But that doesn't make it any less of a standard if there are multiple interoperable implementations.  Also, Safari implemented WebRTC, so I fail to see how their position would make Safari users safer.  Maybe that's just a case of [this principle](https://w3ctag.github.io/design-principles/#leave-the-web-better), which says that you don't double-down on bad web features.  Either way, refusal to implement is only a marginally useful signal in the absence of better information about why.

**Comment by martenrichter** [comment]
I also have the impression, that the usecase (not local test case) itself may be behind the objection from webkit. Especially the advantages of serverCertificateHashes of easier enabling small decentral setups, that do not require larger company infrastructures. As such I think serverCertificateHashes is a ingredient for easier transfer of free knowledge. As such it may hurt business strategies of a gatekeeper.
So I do also not think that this is a discussion about technical issues.

**Comment by alvestrand** [comment]
I would argue that Con 1 (shared private key) is not stated correctly, and should actually be a Pro 5 - the WebRTC usage pattern of generating a private key on-demand, pushing its hash through a channel secured by other means, and verifying the hash during (D)TLS setup, actually gives developers an useful pattern to follow that allows setting up an encrypted connection without needing to move a private key between devices.

**Comment by jan-ivar** [comment]
Meeting:
- waiting on input from webkit.
- Current status is we're lacking consensus to remove.

**Comment by annevk** [comment]
If we need consensus, it seems we would also lack consensus to advance the draft in its current form? It seems weird if there's a high bar for removal if there was not an equivalent high bar for adding.

**Comment by jan-ivar** [comment]
> If we need consensus, it seems we would also lack consensus to advance the draft in its current form? It seems weird if there's a high bar for removal if there was not an equivalent high bar for adding.

For bookkeeping, [serverCertificateFingerprints](https://www.w3.org/TR/2021/WD-webtransport-20210504/#dom-webtransportoptions-servercertificatefingerprints) was in FPWD whose [consensus was affirmed by CfC](https://lists.w3.org/Archives/Public/public-webtransport/2021May/0000.html) on April 14 2021, without objections.

Our [charter](https://www.w3.org/2024/09/webtransport-wg-charter.html#decisions) states that _"All decisions made by the group should be considered resolved unless and until new information becomes available or unless reopened at the discretion of the Chairs."_

Is there new information? Regarding advancement, we've prepared for wide review with https://github.com/w3c/webtransport/pull/641.

**Comment by kixelated** [comment]
> > Hard-coding the server's DTLS certificate hash is not viable because the client generates random credentials too.
> 
> In WebRTC, you can just let the server ignore the client's certificate altogether to simulate regular TLS semantics.

True, but the server needs the client's (random) ICE credentials too. Otherwise STUN messages will fail auth. Unless you can munge those in the local SDP?

**Comment by martenrichter** [comment]
> But for WebTransport, only the server requires a certificate. It's trivial (and tempting) to hard-code a certificate hash in your application to avoid this exchange. I've already had a user of my library attempt to do this, distributing their LetsEncrypt certificate alongside their (open source) server. The justification was easier local development of course.

But this will last for only 14 days. So there is no way to avoid a new certificate, if you do not want your users to update every two weeks.

**Comment by marten-seemann** [comment]
@martenrichter is right, the 2-week limit is a pretty strong forcing function to "do things the right way". 

I deployed WebTransport at my previous job, and it did force us to properly implement certificate rotation. In my experience, being able to hard-code a certificate is only useful for testing and prototyping (but there it actually is a very welcome shortcut!).

Maybe we can even shorten this period to 7 days if we think that 14 days is too much of an incentive to hard-code the certificate for applications that are updated frequently?

**Comment by kixelated** [comment]
> > But for WebTransport, only the server requires a certificate. It's trivial (and tempting) to hard-code a certificate hash in your application to avoid this exchange. I've already had a user of my library attempt to do this, distributing their LetsEncrypt certificate alongside their (open source) server. The justification was easier local development of course.
> 
> But this will last for only 14 days. So there is no way to avoid a new certificate, if you do not want your users to update every two weeks.

Ah true, ignore me. You could pre-emptively generate and rotate them but it's a lot of work.

**Comment by martenrichter** [comment]
> Ah true, ignore me. You could pre-emptively generate and rotate them but it's a lot of work.

This is possible, but the amount of work will hopefully prevent people from pre-generating certificates. One could prevent this if a timestamp in the certificate signed by a trusted entity is included (Similar to the current newspaper in crime dramas on the photo of hostages or https://en.wikipedia.org/wiki/Trusted_timestamping).

**Comment by jan-ivar** [comment]
An informal meeting was held with core implementers on this thread, to clarify the arguments for and against, and to explore alternatives.

At the highest level, there were acknowledged benefits and acknowledged risks with the serverCertificateHashes API.

The key question is whether the former justifies the latter.

Risks:
- Loss of transparency about the endpoints that a service presents.  (Mitigating factor: the site that provides the serverCertificateHashes parameter is responsible; which is an imperfect mitigation.)

Benefit:
- Ephemeral VMs can use this without waiting for certificate issuance.  (Mitigating factor: they still can get real certificates, modulo issuance delays, so this is "just" a performance question.)
- WebRTC works this way and it aids migration if the feature remains.  (Mitigating factor: it's not a great feature in WebRTC either, though we have no path to fixing that in WebRTC.)

A wash:
- Helpful for securing local services for development and IoT; but we really want a better solution for these that is larger and this isn't enough justification on its own to justify this.

Other notes:
- With respect to ephemeral VMs case: if you require public CA, your service now works only as long as CA's issuance service works
- Cert transparency is important to keep track of issuers, but in the local networks use case transparency might be incompatible with privacy
- CSP keyword might be better worded as "unsafe-servercertificate" 

Next steps:
- @annevk to gather feedback from webkit folks not present.

**Comment by alvestrand** [comment]
> A wash:

> Helpful for securing local services for development and IoT; but we really want a better solution for these that is larger and this isn't enough justification on its own to justify this.

Note: My experience (both in IETF and W3C) of waiting for larger and hairier problems to be solved somewhere else before you ship is thoroughly negative; the 2-year delay between completion of the RTCWEB work and the publication of the RFCs was largely driven by other people having to clean up their stuff - and that wasn't even large hairy problems, it was finishing up their documents.

I'd move this firmly to the "positive" column.

**Comment by bergstroem** [comment]
If I may offer our perspective at Vindral, as early users of the API.

We have several customer deployments with varied local network installations that rely on serverCertificateHashes. In many of these cases, removing this feature would either significantly complicate implementation or make it unfeasible.

We fully understand the concerns raised in this discussion. That said, from our perspective, further restricting the use of serverCertificateHashes would be preferable to removing it entirely.

**Comment by alvestrand** [comment]
An interesting aspect of the WebRTC API for certificates is that there is no API for installing an RTCCertificate, no API for extracting the private key from an RTCCertificate, and no way of getting an usable fingerprint except from an RTCCertificate.

RTCCertificates are transferable and storable (in the JS sense), but the keying material is not exposed in any useful fashion, so it's not possible to move an RTCCertificate to another device.

So in WebRTC, sharing private keys among devices using JS interfaces does not exist.

In 10+ years of deploying WebRTC, there hasn't been enough demand for extracting certificates or installing "foreign" certificates to change this.

**Comment by ekinnear** [comment]
Chatting a bit more with @annevk and other WebKit folks in the context of the above risks and benefits, we don't have a lot of affinity for some of the use cases, but we recognize that they hold value for a substantial number of folks, and the potential better answers aren't available today. Since this isn't actively harmful, let's keep it in the spec, and we have no objection to implementing, as well.

From that perspective, we'd like to recommend closing this issue with no action -- keeping `serverCertificateHashes` in the spec. 

Also, I'd like to say many many thanks to everyone who was willing to dig into this so deeply and really analyze it, we've added some great protections as a result of those conversations.

**Comment by jan-ivar** [comment]
Thanks!
